modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/mvcc/MvccDeadlockDetectionConfigTest.java

```diff
@@ -0,0 +1,158 @@
+    /** */
+    private boolean deadlockDetectionEnabled;
+
+    /** */
+    @After
+    public void stopCluster() {
+        stopAllGrids();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        int timeout = deadlockDetectionEnabled ? 1 : 0;
+
+        return super.getConfiguration(igniteInstanceName)
+            .setTransactionConfiguration(new TransactionConfiguration().setDeadlockTimeout(timeout));
+    }
+
+    /** */
+    @Test
+    public void deadlockDetectionDisabled() throws Exception {
+        deadlockDetectionEnabled = false;
+
+        Ignite ign = startGrid();
+
+        IgniteCache<Object, Object> cache = ign.createCache(new CacheConfiguration<>(DEFAULT_CACHE_NAME)
+            .setAtomicityMode(TRANSACTIONAL_SNAPSHOT));
+
+        CyclicBarrier b = new CyclicBarrier(2);
+
+        int txTimeout = 3_000;
+
+        IgniteInternalFuture<?> futA = GridTestUtils.runAsync(() -> {
+            try (Transaction tx = ign.transactions().txStart(PESSIMISTIC, REPEATABLE_READ, txTimeout, 0)) {
+                cache.put(1, 'a');
+                b.await();
+                cache.put(2, 'a');
+            }
+
+            return null;
+        });
+
+        IgniteInternalFuture<?> futB = GridTestUtils.runAsync(() -> {
+            try (Transaction tx = ign.transactions().txStart(PESSIMISTIC, REPEATABLE_READ, txTimeout, 0)) {
+                cache.put(2, 'b');
+                b.await();
+                cache.put(1, 'b');
+            }
+
+            return null;
+        });
+
+        IgniteCheckedException e = awaitCompletion(futA, futB);
+
+        assertTrue(e.toString(), e.hasCause(IgniteTxTimeoutCheckedException.class));
+    }
+
+    /** */
+    @Test
+    public void deadlockDetectionEnabled() throws Exception {
+        deadlockDetectionEnabled = true;
+
+        Ignite ign = startGrid();
+
+        IgniteCache<Object, Object> cache = ign.createCache(new CacheConfiguration<>(DEFAULT_CACHE_NAME)
+            .setAtomicityMode(TRANSACTIONAL_SNAPSHOT));
+
+        CyclicBarrier b = new CyclicBarrier(2);
+
+        IgniteInternalFuture<?> futA = GridTestUtils.runAsync(() -> {
+            try (Transaction tx = ign.transactions().txStart(PESSIMISTIC, REPEATABLE_READ)) {
+                cache.put(1, 'a');
+                b.await();
+                cache.put(2, 'a');
+            }
+
+            return null;
+        });
+
+        IgniteInternalFuture<?> futB = GridTestUtils.runAsync(() -> {
+            try (Transaction tx = ign.transactions().txStart(PESSIMISTIC, REPEATABLE_READ)) {
+                cache.put(2, 'b');
+                b.await();
+                cache.put(1, 'b');
+            }
+
+            return null;
+        });
+
+        IgniteCheckedException e = awaitCompletion(futA, futB);
+
+        assertTrue(e.toString(), e.hasCause(IgniteTxRollbackCheckedException.class));
```

Comment:
Should we check if any DeadLock exception occurs here?

Reply:
Done.